Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,381 @@
+    if (nodes.size() < 1)
         return std::nullopt;
+    
+    StringBuilder searchBufferBuilder;
+    for (auto node : nodes)
+        searchBufferBuilder.append(downcast<Text>(*node).data());
+    searchBuffer = searchBufferBuilder.toString();
+    
+    unsigned searchStart = 0;
+    
+    if (nodes[0] == &searchRange.startContainer())
+        searchStart = searchRange.startOffset();
+    
+    std::optional<unsigned> matchIndex;
+    
+    while (!matchIndex) {
+        // FIXME: find only using the base characters i.e. also fold accents and others.
+        auto potentialIndex = searchBuffer.findIgnoringASCIICase(query, searchStart);
 
-    auto searchRange = makeRangeSelectingNodeContents(document);
+        if (potentialIndex == notFound)
+            return std::nullopt;
+        matchIndex = potentialIndex;
+        
+        start = boundaryPointAtIndexInNodes(*matchIndex, nodes, BoundaryPointIsAtEnd::No);
+        end = boundaryPointAtIndexInNodes(*matchIndex + query.length(), nodes, BoundaryPointIsAtEnd::Yes);
+        
+        if (((wordStartBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex)) || ((wordEndBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex + query.length()))) {
+            searchStart = *matchIndex + 1;
+            matchIndex = std::nullopt;
+        }
+    }
+        
+    unsigned endInset = 0;
+    if (nodes.last() == &searchRange.endContainer())
+        endInset = searchRange.endContainer().length() - searchRange.endOffset();
+    
+    if (*matchIndex + query.length() > searchBuffer.length() - endInset)
+        return std::nullopt;
+    
+    ASSERT_WITH_MESSAGE(start, "Scroll To Text Fragment: Start cannot be null");
+    ASSERT_WITH_MESSAGE(end, "Scroll To Text Fragment: End cannot be null");
+    
+    ASSERT_WITH_MESSAGE(contains(searchRange, start.value()), "Scroll To Text Fragment: Start is within SearchRange.");
+    ASSERT_WITH_MESSAGE(contains(searchRange, end.value()), "Scroll To Text Fragment: End is within SearchRange.");
+    
+    if (!start || !end || !contains(searchRange, start.value()) || !contains(searchRange, end.value()))
+        return std::nullopt;
     
-    FindOptions findOptions { DoNotRevealSelection };
+    return SimpleRange(*start, *end);
     
-    auto resultRange = collapseIfRootsDiffer(findPlainText(searchRange, target, findOptions));
+}
+// https://wicg.github.io/scroll-to-text-fragment/#find-a-string-in-range
+static std::optional<SimpleRange> rangeOfStringInRange(const String& query, SimpleRange& searchRange, WordBounded wordStartBounded, WordBounded wordEndBounded)
+{
+    while (!searchRange.collapsed()) {
+        Node* currentNode = &searchRange.startContainer();
+        
+        if (isNonSearchableSubtree(*currentNode)) {
+            if (auto newStart = nextSkippingChildrenInComposedTreeIgnoringUserAgentShadow(*currentNode))
+                searchRange.start = { *newStart, 0 };
+            else
```

Comment:
Instead of adding else return everywhere.
Why don't we make `searchRange` `optional<SimpleRange>`?
And just check `searchRange && searchRange->collapsed()` as the loop invariant?

Reply:
This only removed two returns and starts to deviate from the shape of the spec, I'd rather not.